### PR TITLE
Interface testing

### PR DESCRIPTION
**Pull request by BrandonArp**
This should look more like what we talked about as far as the interfaces go.

### CODE REVIEW

File: src/main/java/com/arpnetworking/metrics/incubator/impl/TsdPeriodicMetrics.java [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+import javax.annotation.Nullable;
+
+/**
+ * Implementation of a {@link PeriodicMetrics} backed by a {@link TsdMetricsFactory}.
+ *
+ * NOTE: This class must be scheduled with an executor in order for metrics to be recorded.
+ *
+ * @author Brandon Arp (brandon dot arp at inscopemetrics dot com)
+ */
+public final class TsdPeriodicMetrics implements PeriodicMetrics, Runnable {
```
Comment (vjkoskela) [inline src/main/java/com/arpnetworking/metrics/incubator/impl/TsdPeriodicMetrics.java]: The flush method is on this implementation, which will used directly by clients. Those clients should not know that this is a Runnable. Suggestion: use an internal nested class instance.

Comment (vjkoskela) [inline src/main/java/com/arpnetworking/metrics/incubator/impl/TsdPeriodicMetrics.java]: ^ Discussion: Should the Runnable aspect be separated from the TsdPeriodicMetrics itself?

File: src/main/java/com/arpnetworking/metrics/incubator/PeriodicMetrics.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * Set the timer to the specified value. This is most commonly used to
+     * record timers from external sources that are not directly integrated with
+     * metrics.
+     *
+     * @param name The name of the timer.
+     * @param duration The duration of the timer.
+     * @param unit The time unit of the timer.
+     */
+    void recordTimer(String name, long duration, @Nullable TimeUnit unit);
```
Comment (BrandonArp) [inline src/main/java/com/arpnetworking/metrics/incubator/PeriodicMetrics.java]: Done.

File: src/main/java/com/arpnetworking/metrics/incubator/impl/SafeSharedReference.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * A class to ease the acquisition and use of an atomic ref, protected by a ReadWriteLock.
+ * The readLocked and writeLocked methods will perform the passed action after obtaining
+ * a read lock and write lock, respectively.
+ *
+ * @param <T> The type of object protected by the locks.
+ *
+ * @author Brandon Arp (brandon dot arp at inscopemetrics dot com)
+ */
+public final class SafeSharedReference<T> {
```
Comment (BrandonArp) [inline src/main/java/com/arpnetworking/metrics/incubator/impl/SafeSharedReference.java]: I shortened (slightly) to ReadWriteLockedReference.

**Comment by BrandonArp** [comment]
Addressed as discussed, rebased, and merging.

**Pull request merged**